lib/puppet/indirector/facts/facter.rb

```diff
@@ -21,6 +21,13 @@ def save(facts)
   # Lookup a host's facts up in Facter.
   def find(request)
     Facter.reset
+
+    # Add the puppetversion fact; this is done before generating the hash so it is
+    # accessible to custom facts.
+    Facter.add(:puppetversion) do
+      setcode { Puppet.version.to_s }
+    end
+
```

Comment:
We're capturing `puppetversion` here, but `Puppet::Node::Facts#add_local_facts`, called below, also captures the [`clientversion`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/node/facts.rb#L31) fact (which has the same value as `puppetversion`).

```
$ be puppet apply -e 'notice($clientversion)'
Notice: Scope(Class[main]): 4.0.0
```

It would be good to consolidate the two code paths. On one hand, we could just add `values[:puppetversion] = Puppet.version.to_s` to the `add_local_facts` and be done.

On the other hand, if we want these facts to be visible to custom facts, then we have to set them before `Facter.to_hash` is called. That would take more work to unwind, e.g. the network_device also has logic for constructing facts objects: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/facts/network_device.rb#L13-L17

Also we'll want to update the docs about Puppet Agent Facts: https://docs.puppetlabs.com/puppet/latest/reference/lang_facts_and_builtin_vars.html#puppet-agent-facts